design/backup-performance-improvements.md

```diff
@@ -0,0 +1,212 @@
+}
+```
+
+To support this new rpc method, we define new request/response message types:
+```
+message BackupItemActionAdditionalItemsRequest {
+    string plugin = 1;
+    bytes item = 2;
+    bytes backup = 3;
+}
+
+message BackupItemActionAdditionalItemsResponse {
+    repeated generated.ResourceIdentifier additionalItems = 1;
+}
+```
+
+A new PluginKind, `BackupItemActionV3`, will be created, and the backup process will be modified to use this plugin kind. Unlike with the V1->V2 transition, however, we will not provide a V3 adapter for BIAv2, because there is no reliable way to know what items to return from the new method. Instead, Velero will only invoke the new `GetAdditionalItems` method if *all* registered plugins are V3. If there are any V2 plugins registered, then we will continue to use the BIAv2 API, and process item blocks with only one item. If there are both V2 and V3 plugins, we will adapt (i.e. downgrade) V3 plugins to V2 rather than vice versa. In order for Velero to support item block processing out of the box, all existing V1 and V2 plugins in velero core and supported plugins will need to be converted to V3 as part of implementation. If we are adapting to V2 plugins, then velero will *not* call the new V3 plugin API method, and every item returned from the Item Collector will be in its own ItemBlock of one.
+
+### Changes to processing item list from the Item Collector
+
+#### New structs ItemBlock and ItemBlockItem
+```
+type ItemBlock struct {
+    log           logrus.FieldLogger
+    itemBackupper *itemBackupper
+    Items         []ItemBlockItem
+}
+
+type ItemBlockItem struct {
+    gr           schema.GroupResource
+    item         *unstructured.Unstructured
+    preferredGVR schema.GroupVersionResource
+}
+```
+
+#### Current workflow
+In the `BackupWithResolvers` func, the current Velero implementation iterates over the list of items for backup returned by the Item Collector. For each item, Velero loads the item from the file created by the Item Collector, we call `backupItem`, update the GR map if successful, remove the (temporary) file containing item metadata, and update progress for the backup.
+
+#### Modifications to the loop over ItemCollector results
+The current workflow within each iteration of the ItemCollector.items loop will replaced with the following:
+- (note that some of the below should be pulled out into a helper func to facilitate recursive call to it for items returned from `GetAdditonalItems`.)
+- Before loop iteration, create a new `itemsInBlock` map of type map[velero.ResourceIdentifier]bool which represents the set of items already included in a block.
+- If `item` is already in `itemsInBlock`, continue. This one has already been processed.
+- Add `item` to `itemsInBlock`.
+- Load item from ItemCollector file. Close/remove file after loading (on error return or not, possibly with similar anonymous func to current impl)
```

Comment:
This is worth considering. Let me go back and look at the structs again, but it may be that instead of passing full unstructured content to the goroutine we should instead pass the struct from the item collector which contains the file path instead.

Reply:
Actually, in thinking further around this, there's a tradeoff here. We already need to load the full item content from file in order to call GetAdditionalItems, since often plugin logic on what items are needed is based on item content and not just type/name. So if we pass in file path instead, we'll end up loading every item from file twice. Also, since we're using a buffered channel to communicate with the workers, the number of blocks in memory at any given time will be limited. For example, if worker count is 10 and buffer size is 10, then at most 20 blocks will be in memory at a time, even if the backup has thousands of items.